aten/src/ATen/TensorMeta.h

```diff
@@ -10,28 +10,54 @@ class Tensor;
 
 namespace impl {
 
-struct MetaBase {
+// Use this to define the prototype for a meta function.  There are two
+// versions; one that takes one argument (just the operator name), or FUNC2
+// variant that takes two arguments (operator name and overload name).
+//
+// Example usage:
+//
+//    TORCH_META_FUNC2(add, Tensor) (
+//      const Tensor& self, const Tensor& other
+//    ) {
+//      ... compute sizes and options ...
+//      set_output(sizes, options);
+//    }
+//
+#define TORCH_META_FUNC(name) void name::meta
+#define TORCH_META_FUNC2(name, overload) void name##_##overload::meta
+
+// Use this to define the prototype for an implementation.  This takes only
+// one argument, which is the name of the dispatch key entry you're
+// implementing.
+//
+// Example usage:
+//
+//    TORCH_IMPL_FUNC(add_cpu) (
+//      Tensor& result, const Tensor& self, const Tensor& other
+//    ) {
+//      ... do the actual implementation ...
+//    }
+//
+#define TORCH_IMPL_FUNC(name) void structured_##name::impl
```

Comment:
FYI noted in the RFC how it can be weird for a "FUNC" to have stuff like `*this` in it. I might consider a name that doesn't explicitly invite the misunderstanding that it's a bare function... though a confused reader will find a good answer here, so not a strong preference.

Reply:
`TORCH_IMPL_METHOD`?